src/main/java/net/pms/dlna/SubSelFile.java

```diff
@@ -1,98 +1,257 @@
+			}
+			liveResource.setMediaSubtitle(subtitle);
+			liveResource.resetSubtitlesStatus();
+			addChild(liveResource);
+		}
+	}
+
+	private static int removeLanguage(ArrayList<SubtitleItem> subtitles, LinkedHashMap<String, Integer> languages, String languageCode, int count) {
+		ListIterator<SubtitleItem> iterator = subtitles.listIterator(subtitles.size());
+		int removed = 0;
+		while (removed != count && iterator.hasPrevious()) {
+			SubtitleItem subtitle = iterator.previous();
+			if (
+				(
+					languageCode == null &&
+					subtitle.getLanguageCode() == null
+				) ||
+				(
+					languageCode != null &&
+					languageCode.equals(subtitle.getLanguageCode())
+				)
+			) {
+				Integer langCount = languages.get(languageCode);
+				if (langCount != null) {
+					if (langCount.intValue() == 1) {
+						languages.remove(languageCode);
+					} else {
+						languages.put(languageCode, langCount.intValue() - 1);
+					}
+				}
+				iterator.remove();
+				removed++;
+			}
+		}
+		return removed;
+	}
+
+	private static void reduceSubtitles(ArrayList<SubtitleItem> subtitles, int limit) {
+		int remove = subtitles.size() - limit;
+		if (remove <= 0) {
 			return;
 		}
-		List<String> sortedKeys = new ArrayList<>(data.keySet());
-		Collections.sort(sortedKeys, new SubSort(getDefaultRenderer()));
-		for (String key : sortedKeys) {
-			LOGGER.debug("Add play subtitle child " + key + " rf " + orig);
-			DLNAMediaSubtitle sub = orig.getMediaSubtitle();
-			if (sub == null) {
-				sub = new DLNAMediaSubtitle();
+
+		LinkedHashMap<String, Integer> languages = new LinkedHashMap<>();
+		for (SubtitleItem subtitle : subtitles) {
+			String languageCode = isBlank(subtitle.getLanguageCode()) ? null : subtitle.getLanguageCode();
+			if (!languages.containsKey(languageCode)) {
+				languages.put(languageCode, 1);
+			} else {
+				languages.put(languageCode, languages.get(languageCode).intValue() + 1);
 			}
-			String lang = OpenSubtitle.getLang(key);
-			String name = OpenSubtitle.getName(key);
-			sub.setType(SubtitleType.SUBRIP);
-			sub.setId(101);
-			sub.setLang(lang);
-			sub.setLiveSub((String) data.get(key), OpenSubtitle.subFile(name + "_" + lang));
-			DLNAResource nrf = orig.clone();
-			nrf.setMediaSubtitle(sub);
-			nrf.resetSubtitlesStatus();
-			addChild(nrf);
-			if (rf != null) {
-				((RealFile) nrf).ignoreThumbHandling();
+		}
+
+		// Remove those without a specified language first
+		if (languages.containsKey(null)) {
+			remove -= removeLanguage(subtitles, languages, null, remove);
+		}
+
+		if (remove > 0) {
+			// Build a sorted map where each language gets 30 extra points per step in the priority
+			ArrayList<LanguageRankedItem> languageRankedSubtitles = new ArrayList<>();
+			int languagePreference = 0;
+			String languageCode = null;
+			for (SubtitleItem subtitle : subtitles) {
+				if (languageCode == null) {
+					languageCode = subtitle.getLanguageCode();
+					languagePreference = languages.size() * 30;
```

Comment:
How was chosen this value ? 30

Reply:
It's chosen relative to the other values that make up the score. Think of it as "weights" that make out the total score for how good a fit the subtitles are. If you look at the size of the other "weights", you will see how much preference is given to the subtitles because the language is prioritized by the user.

I've set all the weights by estimate only, and the values might have to be tweaked to get the optimal behaviour. Based on my tests the current values seem to work reasonably well as a starting point, but I'm sure you can find cases where it doesn't work that well.